src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/VSRestoreSettingsUtilities.cs

```diff
@@ -0,0 +1,128 @@
+                if (RestoreAdditionalProjectSources.Equals(source.Source))
+                {
+                    readingAdditionalSources = true;
+                }
+                else
+                {
+                    if (readingAdditionalSources)
+                    {
+                        additionalSources.Add(source.Source);
+                    }
+                    else
+                    {
+                        sources.Add(source.Source);
+                    }
+                }
+            }
+
+            var processedSources = (
+                        ShouldReadFromSettings(sources) ?
+                            SettingsUtility.GetEnabledSources(settings).Select(e => e.Source) :
+                            HandleClear(sources))
+                        .Concat(additionalSources);
+
+            // Resolve relative paths
+            return processedSources.Select(e => new PackageSource(
+                UriUtility.GetAbsolutePathFromFile(
+                    sourceFile: project.RestoreMetadata.ProjectPath,
+                    path: e)))
+                .ToList();
+        }
+
+        public static List<string> GetFallbackFolders(ISettings settings, PackageSpec project)
+        {
+            var fallbackFolders = new List<string>();
+            var additionalFallbackFolders = new List<string>();
+
+            var readingAdditionalFallbackFolders = false;
+            foreach (var fallbackFolder in project.RestoreMetadata.FallbackFolders)
+            {
+                if (RestoreAdditionalProjectFallbackFolders.Equals(fallbackFolder))
+                {
+                    readingAdditionalFallbackFolders = true;
+                }
+                else
+                {
+                    if (readingAdditionalFallbackFolders)
+                    {
+                        additionalFallbackFolders.Add(fallbackFolder);
+                    }
+                    else
+                    {
+                        fallbackFolders.Add(fallbackFolder);
+                    }
+                }
+            }
+
+            var processedFallbackFolders = (
+                        ShouldReadFromSettings(fallbackFolders) ?
+                            SettingsUtility.GetFallbackPackageFolders(settings) :
+                            HandleClear(fallbackFolders))
+                        .Concat(additionalFallbackFolders);
+
+
+            // Resolve relative paths
+            return processedFallbackFolders.Select(e =>
+                UriUtility.GetAbsolutePathFromFile(
+                    sourceFile: project.RestoreMetadata.ProjectPath,
+                    path: e))
+                .ToList();
+        }
+
+        private static bool ShouldReadFromSettings(IEnumerable<string> values)
+        {
+            return !values.Any() && values.All(e => !StringComparer.OrdinalIgnoreCase.Equals("CLEAR", e));
+        }
+
+        public static IEnumerable<string> HandleClear(IEnumerable<string> values)
```

Comment:
I'm writing the tests for these functions explicitly as mentioned in the PR. 
The GetSources/GetFallbackFolders ones are tested in integration with the VSSolutionRestoreService